### PR TITLE
Enforce global lock in oci hooks

### PR DESCRIPTION
**Pull request by apostasie**
See
#3518
#2908

and probably others.

At this point, nerdctl simply cannot be used in any non trivial scenario involving CNI and concurrency.

Besides local context, this started cropping up as soon as I disabled retries on the CI and cranked up parallelism on tests.

The bottom-line MAY be that CNI plugins cannot be used concurrently at all - whether by design or not - this is unconfirmed though.

A "proper" fix is likely very involved - probably rewriting most / all of CNIEnv, or the cni plugins themselves. As floated in the tickets, it is not even clear if separate binaries + ocihooks is a tenable design at all.

Anyhow, this is the best I can think of right now that does not involve too much work: enforce a global lock in hooks. It is a hammer. It will hurt performance, likely. But then the choice is "does not work" vs. "works slowly".
As stated in the comment, it will probably not even cover all cases - but it is better than nothing and will make things a bit more acceptable.

This has been forked out of #3517 which is currently ground zero.

### CODE REVIEW

**Comment by apostasie** [comment]
@AkihiroSuda at your convenience.
(pending green)

**Comment by apostasie** [comment]
Failure is the usual TestIPFSComposeUp/stargz

File: pkg/ocihook/ocihook.go [hunk 1]
```diff
@@ -92,6 +93,27 @@ func Run(stdin io.Reader, stderr io.Writer, event, dataStore, cniPath, cniNetcon
...
+	// See
+	// https://github.com/containerd/nerdctl/issues/3518
+	// https://github.com/containerd/nerdctl/issues/2908
+	// and likely others
+	// Fixing these issues would require a lot of work, possibly even stopping using individual cni binaries altogether
+	// or at least being very mindful in what operation we call inside CNIEnv at what point, with filesystem locking.
+	// This below is a stopgap solution that just enforces a global lock
+	// Note this here is probably not enough, as concurrent CNI operations may happen outside of the scope of ocihooks
+	// through explicit calls to Remove, etc.
+	lockDir := filepath.Join(cniNetconfPath)
```
Comment (AkihiroSuda) [inline pkg/ocihook/ocihook.go:106]: Doesn't seem to be joined

Comment (apostasie) [reply pkg/ocihook/ocihook.go:106]: Geez. 🤦🏻‍♂️🐟
on my phone.
Will fix.

File: pkg/ocihook/ocihook.go [hunk 2]
```diff
@@ -92,6 +93,26 @@ func Run(stdin io.Reader, stderr io.Writer, event, dataStore, cniPath, cniNetcon
...
+	// See
+	// https://github.com/containerd/nerdctl/issues/3518
+	// https://github.com/containerd/nerdctl/issues/2908
+	// and likely others
+	// Fixing these issues would require a lot of work, possibly even stopping using individual cni binaries altogether
+	// or at least being very mindful in what operation we call inside CNIEnv at what point, with filesystem locking.
+	// This below is a stopgap solution that just enforces a global lock
+	// Note this here is probably not enough, as concurrent CNI operations may happen outside of the scope of ocihooks
+	// through explicit calls to Remove, etc.
+	err = os.MkdirAll(cniNetconfPath, 0o777)
```
Comment (AkihiroSuda) [inline pkg/ocihook/ocihook.go:106]: Why 777?

Comment (apostasie) [reply pkg/ocihook/ocihook.go:106]: Good catch.
I don't think there is a reason, and it should work with 0o700

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**